pkg/network/handlers/work_package_submission.go

```diff
@@ -0,0 +1,129 @@
+package handlers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/quic-go/quic-go"
+
+	"github.com/eigerco/strawberry/internal/crypto"
+	"github.com/eigerco/strawberry/internal/work"
+	"github.com/eigerco/strawberry/pkg/serialization/codec/jam"
+)
+
+// ImportedSegmentsFetcher defines an interface for fetching imported segments from the availability system
+type ImportedSegmentsFetcher interface {
+	FetchImportedSegment(hash crypto.Hash) ([]byte, error)
+}
+
+// ImportSegments implements ImportedSegmentsFetcher
+type ImportSegments struct{}
+
+func (m *ImportSegments) FetchImportedSegment(hash crypto.Hash) ([]byte, error) {
+	fmt.Printf("fetching imported segment for hash %x\n", hash)
+	// TODO implement
+
+	return []byte{}, nil
+}
+
+// WorkPackageSubmissionHandler processes incoming CE-133 submission streams
+type WorkPackageSubmissionHandler struct {
+	Fetcher ImportedSegmentsFetcher
+}
+
+// NewWorkPackageSubmissionHandler creates a new handler instance with the given fetcher.
+func NewWorkPackageSubmissionHandler(fetcher ImportedSegmentsFetcher) *WorkPackageSubmissionHandler {
+	return &WorkPackageSubmissionHandler{
+		Fetcher: fetcher,
+	}
+}
+
+// HandleStream implements the guarantor side of the CE-133 protocol.
+// It expects two messages:
+//
+//	Message 1: [Core Index (u16) ++ work.Package]
+//	Message 2: [Extrinsic data (raw bytes)]
+//
+// After reading these it should fetch imported segments from the availability system
+func (h *WorkPackageSubmissionHandler) HandleStream(ctx context.Context, stream quic.Stream) error {
+	msg1, err := ReadMessageWithContext(ctx, stream)
+	if err != nil {
+		return fmt.Errorf("failed to read message 1: %w", err)
+	}
+
+	var coreIndex uint16
+	if err = jam.Unmarshal(msg1.Content[:2], &coreIndex); err != nil {
+		return fmt.Errorf("failed to unmarshal core index: %w", err)
+	}
+
+	var pkg work.Package
+	if err = jam.Unmarshal(msg1.Content[2:], &pkg); err != nil {
```

Comment:
Just a thought, shouldn't we be trying to validate the size of the message before we unmarshal?

Reply:
I added a simple bound check for the message content